packages/apollo-server-plugin-base/src/index.ts

```diff
@@ -49,13 +51,17 @@ export interface ApolloServerPlugin<
   serverWillStart?(
     service: GraphQLServiceContext,
   ): Promise<GraphQLServerListener | void>;
+
   requestDidStart?(
     requestContext: GraphQLRequestContext<TContext>,
   ): Promise<GraphQLRequestListener<TContext> | void>;
 }
 
 export interface GraphQLServerListener {
+  schemaDidLoadOrUpdate?(schemaContext: GraphQLSchemaContext): void;
```

Comment:
In Apollo Server 3, all plugin callbacks are async other than willResolveField (I am paranoid about the performance impact of making that one async, though I did not actually benchmark).

Would it be possible to make this one async too, or would that introduce race conditions?

In any case, the docs (docs/source/integrations/plugins-event-reference.md and perhaps docs/source/integrations/plugins.md) should be updated with the new hook (possibly updating the parts that explain that all but one hook is async).

Reply:
It could be fair keep this one as synchronous to make SchemaManager simpler.

Or it could be async like everything else and just document that there's nothing stopping calls to them from overlapping.  That might be simpler to think about.  (and processSchemaLoadOrUpdateEvent wouldn't await, it would just `.catch` with the same logger we already have.)

In any case, docs need updating.

Reply:
I'd say the main issues with async callbacks here are that they'd indeed increase the complexity of `SchemaManager`, and that it's not necessarily clear what the desired ordering semantics would be for the callbacks, e.g. with questions like:
- For a given schema load/update event, should its callbacks be executed concurrently or in some consistent sequential order?
- For a given schema load/update event, should the competion of its callbacks block subsequent schema updates?
  - If not, then for two schema load/update events, should all the callbacks of the prior event complete before all the callbacks of the subsequent event start?
    - If not that either, then for two schema load/update events, for each callback, should that callback complete for the prior event before the callback is started for the subsequent event?

I'm not necessarily sure what behavior users would want here, so I'd rather figure that out first and build it later.

(Still in the middle of updating the docs.)